### ISSUE TITLE
Cursor movement slow/laggy when semicolon present in Vi mode

### ISSUE THREAD

**Issue by ivangreene** [dbcli/pgcli#994]
## Description

Moving around a multi-line query to edit it is extremely slow+laggy in Vi mode when a semicolon is present at the end of the line. Deleting the semicolon instantly speeds up movement for some reason. This is obnoxious in multiline mode when a semicolon is required to send a query and when one wants to move up to edit the query.
## Your environment

- [x] Please provide your OS and version information.
MacOS 10.14.2
- [x] Please provide your CLI version.
2.0.2
- [x] What is the output of ``pip freeze`` command.
Empty for some reason

**Comment by eeshugerman** [comment]
I'm running into this as well.

OS: GNU/Linux (Solus)
CLI version: 2.02
`pip freeze`:
```
attrs==18.2.0
awscli==1.16.51
boltons==19.1.0
botocore==1.12.41
cli-helpers==1.1.0
Click==7.0
colorama==0.3.9
configobj==5.0.6
docutils==0.14
face==0.1.0
glom==19.2.0
gnupg==2.3.1
humanize==0.5.1
jmespath==0.9.3
pgcli==2.0.2
pgspecial==1.11.5
prompt-toolkit==2.0.8
psutil==5.4.8
psycopg2==2.7.7
pyasn1==0.4.4
pycairo==1.18.0
Pygments==2.3.1
PyGObject==3.30.1
pyopt==0.84
pyotp==2.2.7
python-dateutil==2.7.5
python-gnupg==0.4.3
pytotp==3
PyYAML==3.13
rsa==3.4.2
s3transfer==0.1.13
setproctitle==1.1.10
six==1.11.0
sqlparse==0.2.4
tabulate==0.8.3
terminaltables==3.1.0
urllib3==1.24.1
wcwidth==0.1.7
```

**Comment by j-bennet** [comment]
@jonathanslenders Would you be able to have a look?

**Comment by Ovyerus** [comment]
Getting this too, both in Vi and Emacs mode.

OS: Windows 10 Pro 1803
CLI version: 2.0.2
pip freeze: [gist](https://gist.github.com/Ovyerus/efeb08e0782ce5cd3b50b7b85cb756d8)

**Comment by j-bennet** [comment]
@jonathanslenders I have the same problem in emacs mode as well. It is quite pronounced.

**Comment by jonathanslenders** [comment]
I'm trying to reproduce it, but so far I don't have any issues. Can any of you reproduce it with a local, empty database?

**Comment by Ovyerus** [comment]
Haven't upgraded pgcli since, and I just tested it now on both a new empty db, and the same one as before and it seems to have fixed itself?

**Comment by eeshugerman** [comment]
Weird, I can't reproduce it either right now. If it happens again I'll see if I will make note. I did mess around with my config file recently so maybe it has something to do with that...

**Comment by isobit** [comment]
Having the same issue. Did some cProfiling and by comparing the percall times I was able to trace it down to the `_is_complete` method in `pgbuffer.py` (see [here](https://github.com/dbcli/pgcli/blob/8cb7009bcd0f0062942932c853706a36178f566c/pgcli/pgbuffer.py#L28)). Looks like [`is_open_quote`](https://github.com/dbcli/pgcli/blob/8cb7009bcd0f0062942932c853706a36178f566c/pgcli/packages/parseutils/utils.py#L108) is slow on large inputs, and it's only called when that `and` isn't short-circuited, which is when multi-line mode is on and there's a semicolon at the end of the statement.

Disabling the `and not is_open_quote(sql)` seems to fix the issue on my machine. Maybe there's a way to remove that check or improve the performance of it?

**Comment by owst** [comment]
This should be fixed on master due to #1109 - @isobit's analysis was spot-on - we were unnecessarily calling the expensive `is_open_quote` method (way) too frequently due to the way the multi-line mode was being switched on/off dynamically. Now, we only run that method once, on Enter keypress

**Comment by Ovyerus** [comment]
Did a quick test, and it does seem to not be happening anymore. 🎉

**Comment by j-bennet** [comment]
Fixed in 2.2.0.
